pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/OverloadShedder.java

```diff
@@ -40,73 +44,89 @@
  * rate that has not been recently unloaded.
  */
 public class OverloadShedder implements LoadSheddingStrategy {
+
     private static final Logger log = LoggerFactory.getLogger(OverloadShedder.class);
-    private Map<String, String> selectedBundlesCache;
 
-    /**
-     * Create an OverloadShedder with the service configuration.
-     *
-     * @param conf
-     *            Service configuration to create from.
-     */
-    public OverloadShedder(final ServiceConfiguration conf) {
-        selectedBundlesCache = new HashMap<>();
-    }
+    private final Multimap<String, String> selectedBundlesCache = ArrayListMultimap.create();
+
+    private final static double ADDITIONAL_THRESHOLD_PERCENT_MARGIN = 0.05;
 
     /**
-     * Attempt to shed one bundle off every broker which is overloaded.
+     * Attempt to shed some bundles off every broker which is overloaded.
      *
      * @param loadData
      *            The load data to used to make the unloading decision.
      * @param conf
      *            The service configuration.
      * @return A map from bundles to unload to the brokers on which they are loaded.
      */
-    public Map<String, String> findBundlesForUnloading(final LoadData loadData, final ServiceConfiguration conf) {
+    public Multimap<String, String> findBundlesForUnloading(final LoadData loadData, final ServiceConfiguration conf) {
         selectedBundlesCache.clear();
         final double overloadThreshold = conf.getLoadBalancerBrokerOverloadedThresholdPercentage() / 100.0;
         final Map<String, Long> recentlyUnloadedBundles = loadData.getRecentlyUnloadedBundles();
```

Comment:
Is it OK to avoid shed bundle, if all brokers are overload here?

Reply:
Yes, so we had that check in the older SimpleLoadManagerImpl. Not sure why it was removed in ModularLoadManagerImpl. 

It's worth to note though that there's a grace period for a bundle: after being unloaded it will be not be unloaded again for a certain amount of time.